test/client/storage.test.ts

```diff
@@ -37,4 +43,87 @@ describe("StateMemoryEngine", function() {
         engine.revert(checkpoint);
         assert.equal(engine.root, root1, "revert to previous root");
     });
+
+    describe("findVacantSubtree", () => {
+        const hasher = Hasher.new("bytes", ZERO_BYTES32);
+        const zeroes = hasher.zeros(maxDepth);
+
+        it("finds first vacant subtree when empty", async function() {
+            const { path, witness } = await engine.findVacantSubtree(
+                maxSubtreeDepth
+            );
+            assert.equal(path, 0);
+            assert.lengthOf(witness, 30);
+            assert.equal(
+                computeRoot(zeroes[witness.length], path, witness),
+                engine.root
+            );
+        });
+
+        it("finds next vacant subtree when first is filled", async function() {
+            for (let i = 0; i < 2 ** maxSubtreeDepth; i++) {
+                await engine.update(i, states[i]);
+            }
+            await engine.commit();
+
+            const { path, witness } = await engine.findVacantSubtree(
+                maxSubtreeDepth
+            );
+            assert.equal(path, 1);
+            assert.lengthOf(witness, 30);
+            assert.equal(
+                computeRoot(zeroes[witness.length], path, witness),
+                engine.root
+            );
+        });
+
+        it("finds next vacant subtree when first has cached items", async function() {
+            await engine.update(2, states[2]);
+            await engine.update(5, states[5]);
+
+            const { path, witness } = await engine.findVacantSubtree(
+                maxSubtreeDepth
+            );
+            assert.equal(path, 2);
+            assert.lengthOf(witness, 30);
+            assert.equal(
+                computeRoot(zeroes[witness.length], path, witness),
+                engine.root
+            );
+        });
+
+        it("fails when tree is full", async function() {
+            const depth = 4;
+            const smallTreeEngine = new StateMemoryEngine(depth);
+            for (let i = 0; i < 2 ** depth; i++) {
+                await smallTreeEngine.update(i, State.new(i, i, i, i));
+            }
+            await smallTreeEngine.commit();
+
+            let errMsg = "";
+            try {
+                await smallTreeEngine.findVacantSubtree(maxSubtreeDepth);
+            } catch (err) {
+                errMsg = err.message;
+            }
+            assert.equal(
+                errMsg,
+                `Tree at level ${depth -
+                    maxSubtreeDepth} is full, no room for subtree insert`
+            );
+        });
+    });
+
+    describe("updateBatch", () => {
+        it("updates items at correct itemID", async function() {
+            const subtreeID = 2;
+            const items = states.slice(8, 12);
+
+            await engine.updateBatch(subtreeID, maxSubtreeDepth, items);
+
```

Comment:
nitpicks: Though we are throwing the commit away soon, adding `await engine.commit()` here is more consistent with the tests above.

Reply:
Added